File: hindsight-tools/hindsight-fs/src/daemon.ts
/**
 * Background daemon management. `start` spawns a detached copy of this CLI in
 * `__run` mode (the hidden loop entrypoint), writing a pidfile and a log under
 * the mount's control directory. `stop` signals it; `status` reports liveness.
 */

import { promises as fs } from "node:fs";
import { openSync } from "node:fs";
import { spawn } from "node:child_process";
import { fileURLToPath } from "node:url";
import * as path from "node:path";
import { CONTROL_DIR, PID_FILE, LOG_FILE } from "./paths.js";
import { saveConfig, type MountConfig } from "./config.js";

interface PidRecord {
  pid: number;
  bankId: string;
  intervalSeconds: number;
  startedAt: string;
}

function controlPath(dir: string, file: string): string {
  return path.join(dir, CONTROL_DIR, file);
}

/** True if a process with `pid` is currently alive. */
export function isAlive(pid: number): boolean {
  try {
    process.kill(pid, 0);
    return true;
  } catch (err) {
    return (err as NodeJS.ErrnoException).code === "EPERM";
  }
}

export async function readPidRecord(dir: string): Promise<PidRecord | null> {
  try {
    return JSON.parse(await fs.readFile(controlPath(dir, PID_FILE), "utf8")) as PidRecord;
  } catch {
    return null;
  }
}

async function removePidFile(dir: string): Promise<void> {
  try {
    await fs.unlink(controlPath(dir, PID_FILE));
  } catch {
    /* ignore */
  }
}

/** Resolve the path to this CLI's entry script for re-spawning. */
function cliEntry(): string {
  // dist/daemon.js → dist/cli.js
  return path.join(path.dirname(fileURLToPath(import.meta.url)), "cli.js");
}

export interface StartResult {
  pid: number;
  alreadyRunning: boolean;
}

export async function startDaemon(config: MountConfig): Promise<StartResult> {
  const existing = await readPidRecord(config.dir);
  if (existing && isAlive(existing.pid)) {
    return { pid: existing.pid, alreadyRunning: true };
  }

  await fs.mkdir(path.join(config.dir, CONTROL_DIR), { recursive: true });
  await saveConfig(config);

  const logFd = openSync(controlPath(config.dir, LOG_FILE), "a");
  const child = spawn(process.execPath, [cliEntry(), "__run", "--dir", config.dir], {
    detached: true,
    stdio: ["ignore", logFd, logFd],
    env: process.env,
  });
  child.unref();

  if (child.pid === undefined) {
    throw new Error("Failed to spawn daemon process");
  }

  const record: PidRecord = {
    pid: child.pid,
    bankId: config.bankId,
    intervalSeconds: config.intervalSeconds,
    startedAt: new Date().toISOString(),
  };
  await fs.writeFile(
    controlPath(config.dir, PID_FILE),
    JSON.stringify(record, null, 2) + "\n",
    "utf8"
  );

  return { pid: child.pid, alreadyRunning: false };
}

export interface StopResult {
  stopped: boolean;
  pid: number | null;
}

export async function stopDaemon(dir: string): Promise<StopResult> {
  const record = await readPidRecord(dir);
  if (!record) return { stopped: false, pid: null };

  if (!isAlive(record.pid)) {
    await removePidFile(dir);
    return { stopped: false, pid: record.pid };
  }

  try {
    process.kill(record.pid, "SIGTERM");
  } catch {
    /* may have just exited */
  }

  // Give it a moment to exit, then force-kill if needed.
  for (let i = 0; i < 50 && isAlive(record.pid); i++) {
    await new Promise((r) => setTimeout(r, 100));
  }
  if (isAlive(record.pid)) {
    try {
      process.kill(record.pid, "SIGKILL");
    } catch {
      /* ignore */
    }
  }

  await removePidFile(dir);
  return { stopped: true, pid: record.pid };
}

export interface DaemonStatus {
  running: boolean;
  pid: number | null;
  record: PidRecord | null;
}

export async function daemonStatus(dir: string): Promise<DaemonStatus> {
  const record = await readPidRecord(dir);
  if (!record) return { running: false, pid: null, record: null };
  const running = isAlive(record.pid);
  if (!running) await removePidFile(dir);
  return { running, pid: running ? record.pid : null, record };
}

export function logPath(dir: string): string {
  return controlPath(dir, LOG_FILE);
}

File: hindsight-cli/src/commands/fs/daemon.rs
//! Background daemon management. `start` spawns a detached copy of this CLI in
//! `fs run` mode (the hidden loop entrypoint), writing a pidfile and a log under
//! the mount's control directory. `stop` signals it; `status` reports liveness.
//!
//! POSIX-only: the detached refresh loop relies on `setsid`/`kill`. On non-unix
//! targets the daemon commands return an error rather than pretending to work.

use anyhow::Result;
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::{Path, PathBuf};

use super::config::{save_config, MountConfig};
use super::paths::{CONTROL_DIR, LOG_FILE, PID_FILE};

/// Persisted daemon metadata (`<dir>/.hindsight-fs/daemon.pid`).
#[derive(Debug, Serialize, Deserialize)]
pub struct PidRecord {
    pub pid: i32,
    #[serde(rename = "bankId")]
    pub bank_id: String,
    #[serde(rename = "intervalSeconds")]
    pub interval_seconds: u64,
    #[serde(rename = "startedAt")]
    pub started_at: String,
}

fn control_path(dir: &Path, file: &str) -> PathBuf {
    dir.join(CONTROL_DIR).join(file)
}

pub fn log_path(dir: &Path) -> PathBuf {
    control_path(dir, LOG_FILE)
}

/// True if a process with `pid` is currently alive.
#[cfg(unix)]
pub fn is_alive(pid: i32) -> bool {
    // kill(pid, 0): 0 => alive; EPERM => alive but not ours; ESRCH => gone.
    let rc = unsafe { libc::kill(pid, 0) };
    if rc == 0 {
        return true;
    }
    std::io::Error::last_os_error().raw_os_error() == Some(libc::EPERM)
}
#[cfg(not(unix))]
pub fn is_alive(_pid: i32) -> bool {
    false
}

pub fn read_pid_record(dir: &Path) -> Option<PidRecord> {
    let raw = fs::read_to_string(control_path(dir, PID_FILE)).ok()?;
    serde_json::from_str(&raw).ok()
}

fn remove_pid_file(dir: &Path) {
    let _ = fs::remove_file(control_path(dir, PID_FILE));
}

/// Outcome of a `start` request.
pub struct StartResult {
    pub pid: i32,
    pub already_running: bool,
}

#[cfg(unix)]
pub fn start_daemon(config: &MountConfig) -> Result<StartResult> {
    use std::os::unix::process::CommandExt;
    use std::process::{Command, Stdio};

    if let Some(existing) = read_pid_record(&config.dir) {
        if is_alive(existing.pid) {
            return Ok(StartResult {
                pid: existing.pid,
                already_running: true,
            });
        }
    }

    fs::create_dir_all(config.dir.join(CONTROL_DIR))?;
    save_config(config)?;

    let log = fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(control_path(&config.dir, LOG_FILE))?;
    let log_err = log.try_clone()?;

    let exe = std::env::current_exe()?;
    let mut cmd = Command::new(exe);
    // `dir` is the positional arg on the hidden `fs run` subcommand.
    cmd.arg("fs")
        .arg("run")
        .arg(&config.dir)
        .stdin(Stdio::null())
        .stdout(Stdio::from(log))
        .stderr(Stdio::from(log_err));
    // Detach from the controlling terminal so the loop survives shell exit.
    unsafe {
        cmd.pre_exec(|| {
            libc::setsid();
            Ok(())
        });
    }

    let child = cmd.spawn()?;
    let pid = child.id() as i32;
    // Intentionally do not wait — the child is detached and outlives us.

    let record = PidRecord {
        pid,
        bank_id: config.bank_id.clone(),
        interval_seconds: config.interval_seconds,
        started_at: now_iso(),
    };
    let mut body = serde_json::to_string_pretty(&record)?;
    body.push('\n');
    fs::write(control_path(&config.dir, PID_FILE), body)?;

    Ok(StartResult {
        pid,
        already_running: false,
    })
}
#[cfg(not(unix))]
pub fn start_daemon(_config: &MountConfig) -> Result<StartResult> {
    Err(anyhow::anyhow!(
        "Background daemon mode is only supported on Unix. Use `hindsight fs mount` (foreground) instead."
    ))
}

/// Outcome of a `stop` request.
pub struct StopResult {
    pub stopped: bool,
    pub pid: Option<i32>,
}

#[cfg(unix)]
pub fn stop_daemon(dir: &Path) -> Result<StopResult> {
    let record = match read_pid_record(dir) {
        Some(r) => r,
        None => {
            return Ok(StopResult {
                stopped: false,
                pid: None,
            })
        }
    };

    if !is_alive(record.pid) {
        remove_pid_file(dir);
        return Ok(StopResult {
            stopped: false,
            pid: Some(record.pid),
        });
    }

    unsafe {
        libc::kill(record.pid, libc::SIGTERM);
    }

    // Give it a moment to exit, then force-kill if needed.
    for _ in 0..50 {
        if !is_alive(record.pid) {
            break;
        }
        std::thread::sleep(std::time::Duration::from_millis(100));
    }
    if is_alive(record.pid) {
        unsafe {
            libc::kill(record.pid, libc::SIGKILL);
        }
    }

    remove_pid_file(dir);
    Ok(StopResult {
        stopped: true,
        pid: Some(record.pid),
    })
}
#[cfg(not(unix))]
pub fn stop_daemon(_dir: &Path) -> Result<StopResult> {
    Ok(StopResult {
        stopped: false,
        pid: None,
    })
}

/// Liveness snapshot for `status`/health.
pub struct DaemonStatus {
    pub running: bool,
    pub pid: Option<i32>,
    pub record: Option<PidRecord>,
}

pub fn daemon_status(dir: &Path) -> DaemonStatus {
    match read_pid_record(dir) {
        None => DaemonStatus {
            running: false,
            pid: None,
            record: None,
        },
        Some(record) => {
            let running = is_alive(record.pid);
            if !running {
                remove_pid_file(dir);
            }
            DaemonStatus {
                running,
                pid: if running { Some(record.pid) } else { None },
                record: Some(record),
            }
        }
    }
}

fn now_iso() -> String {
    chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true)
}
